File size checking & log-file reopening

shikharvashistha opened the pull request:
<!--
Thank you for contributing to syslog-ng. Please make sure you:
- Read our Contribution guideline: https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md
- Checked that tests pass (including stylechecks: `make style-check`)
- Wrote a news file, if applicable: https://github.com/syslog-ng/syslog-ng/tree/master/news
-->


<!--
PR description
In the description, please explain the problem your pull request intends to solve, and a give general overview of the implementation.
For more information, see: https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#pr-description
-->
This pull implements the functionality to check file size for rotation & reopening if required.

shikharvashistha committed:
file-rotation: set_date_format

This commit introduces functionality for setting up date format based of last_rotation_time

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: file size checking, rename the file

This commit includes the file size checking and renaming if it exceeds the desired size

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: reopen log file & fixes

This commit introduces few fixes to file size checking and rotation including reopening the log file based of the request

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: Added functional test

This commit adds functional test for file-rotation plugin

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -63,6 +93,26 @@ file_rotation_set_size(FileRotationPlugin *self, gsize size)
+
+  g_free(self->date_format);
+  self->date_format = g_strconcat("", date_format, NULL);
+  printf("%s", self->date_format);
```
Please do not use printf(), syslog-ng is not supposed to output anything on stdout.

msg_debug() and msg_trace() can be used, when needed, but I think this is not the case here.

shikharvashistha replied:
Okay then I'll consider removing it

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -63,6 +93,26 @@ file_rotation_set_size(FileRotationPlugin *self, gsize size)
+    }
+
+  g_free(self->date_format);
+  self->date_format = g_strconcat("", date_format, NULL);
```
You can use `g_strdup()`, if you want to get a newly allocated copy.

shikharvashistha replied:
Yeah that's the case here thanks for the suggestion

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -54,6 +83,7 @@ file_rotation_free(LogDriverPlugin *s)
 void
 file_rotation_set_interval(FileRotationPlugin *self, gchar *interval)
 {
+  g_free(self->interval);
```
A g_strdup() call is needed too, as we take ownership of interval.

shikharvashistha replied:
:+1: updated the same accordingly

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,40 @@
+          struct tm *tm = localtime(&now);
+          gchar *date = g_new0(gchar, 100);
+
+          strftime(date, 100, self->date_format, tm);
```
This configuration crashes:

```
log {
  source { stdin(); };

  destination {
    file("/tmp/test" file-rotation(size(12)));
  };
};
```

The current implementation does not use the `rotate(integer)` option to rotate a certain number of files (which would be a nice default), we only have `date-format()` at the moment, so that option needs to be made mandatory.

You can implement an option validation method, and report errors using `msg_error()` and by returning FALSE from `file_rotation_attach_to_driver()`.

shikharvashistha replied:
- So there's a need of new option which tells how many times a file can be rotated ?

- And an option validation method inside `file_rotation_attach_to_driver()` method ?

MrAnno replied:
I'm not sure we have the time to add the `rotate()` option, but feel free to implement it whenever you have time between review rounds.

With `file-rotation(size(300M), rotate(3))`, the new filename would be originalfilename.1, older files would be originalfilename.2 and originalfilename.3. rotate(3) means that we keep only the 3 latest files and remove the older ones.

In case you can implement this, we should not make `date-format()` mandatory.

In case you don't have time to implement the above functionality, we should make `date-format()` mandatory (by adding a validation check to file_rotation_attach_to_driver()).

shikharvashistha replied:
Yeah I've implemented the same https://imgur.com/a/acd9KiW

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,40 @@
+
+          time_t now = time(NULL);
+          struct tm *tm = localtime(&now);
+          gchar *date = g_new0(gchar, 100);
```
This leaks memory.
I think there is no need for allocating memory on the heap, you can use a buffer on the stack.

shikharvashistha replied:
Updated it with a buffer on stack implementation ```gchar date[100]```

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,40 @@
+          gchar *date = g_new0(gchar, 100);
+
+          strftime(date, 100, self->date_format, tm);
+          data->last_rotation_time = g_new0(gchar, 100);
```
This is unused.

MrAnno commented on modules/affile/file-signals.h:
```diff
@@ -30,7 +30,11 @@ typedef struct _FileFlushSignalData FileFlushSignalData;
 {
-  const gchar *filename;
+  gchar *filename;
+  gchar *persist_name;
```
What is the purpose of this field?

shikharvashistha replied:
I thought of using this field to keep a track of last file name.

MrAnno commented on modules/affile/file-signals.h:
```diff
@@ -30,7 +30,11 @@ typedef struct _FileFlushSignalData FileFlushSignalData;
+  gchar *persist_name;
+  gsize size;
+  gboolean *reopen;
+  gchar *last_rotation_time;
```
This seems to be unused. Do we need it?

shikharvashistha replied:
Basically I used it to check the interval of the last rotation, but we can remove it.

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -244,6 +264,8 @@ log_proto_file_writer_new(LogTransport *transport, const LogProtoClientOptions *
   self->signal_slot_connector = connector;
   self->fd = transport->fd;
+  self->writer = writer;
+  self->filename = g_strdup(filename);
```
This leaks memory.

shikharvashistha replied:
As discussed above I've made filename a const gchar* and have updated the same.

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -135,10 +138,27 @@ log_proto_file_writer_flush(LogProtoClient *s)
   self->sum_len = 0;
 
-  FileFlushSignalData signal_data = {};
+  gchar *file_name = g_strdup_printf("%s", self->filename);
```
This leaks memory.
I think we can just use a const gchar, without duplicating anything.

shikharvashistha replied:
Agree, updated

shikharvashistha committed:
file-signals: add file reopener

Added file reopener struct to file-signal.h

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -135,10 +138,27 @@ log_proto_file_writer_flush(LogProtoClient *s)
 
-  FileFlushSignalData signal_data = {};
+  gchar *file_name = g_strdup_printf("%s", self->filename);
+  gsize file_size = lseek(self->fd, 0, SEEK_END);
```
`lseek` might be a cheaper way to check the file's size, but this will be called unconditionally, even if the file-rotation plugin is not configured.

I think it would be better to query the file size in the file-rotation plugin using fstat or stat.
What do you think?

shikharvashistha replied:
But isn't it a more expensive call?

shikharvashistha replied:
Okay it's a 3 vs 1 syscall's difference

MrAnno replied:
My only concern is that calling lseek here adds an unnecessary cost for the file destination. We calculate the file size, and most of the time, we do nothing with it, because our users do not always use file rotation.

It also looks a bit odd in the file writer because it is hard to explain why a flush signal needs the file's size in general.

shikharvashistha replied:
I've replaced it with fstat implementation

shikharvashistha committed:
affile: allow reopening the given file from the flush signal

This commit adds reopening of a file via a flush signal

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: request file reopening from the destination

This commit adds the functionality to reopen the destination file via reopener

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,34 @@
 {
   msg_trace("File rotation signal received", evt_tag_str("filename", data->filename));
+
+  if(data != NULL)
```
NULL should never happen here.
If we want to make sure that the API is used correctly, we can add `g_assert(data);` instead of this condition.

shikharvashistha replied:
OK yes have missed that point.

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,34 @@
+    {
+      if(data->size >= self->size)
+        {
+          msg_debug("File size reached the limit", evt_tag_str("filename", data->filename));
```
I think we could produce a `msg_info()` here: "File size limit reached, rotating file".
We could add the original filename, the new filename, and the size to the info message as well.

shikharvashistha replied:
Updated Accordingly

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,34 @@
+      if(data->size >= self->size)
+        {
+          msg_debug("File size reached the limit", evt_tag_str("filename", data->filename));
+          msg_debug("Sending file rotation signal", evt_tag_str("filename", data->filename));
```
This is not needed anymore.

shikharvashistha replied:
Removed

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -63,6 +87,26 @@ file_rotation_set_size(FileRotationPlugin *self, gsize size)
+void
+file_rotation_set_date_format(FileRotationPlugin *self, gchar *date_format)
+{
+  if (date_format[0] != '-')
```
I think we should let the user specify any format they want.

shikharvashistha replied:
Removing the conditions then

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -42,22 +41,16 @@ struct _FileRotationPlugin
-  gint64 last_rotation;
-  gint64 last_size_check;
-  gint64 last_interval_check;
+  time_t *last_rotation_time;
```
Are you planning to use this field for something?

shikharvashistha replied:
Thought of using this field to keep a track of last rotated time

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -24,11 +24,45 @@
+
+      g_free(new_filename);
+
+      self->number_of_time_rotated++;
```
As I see you've started working on the `rotate()` option. The implementation of this option is not trivial: we need to check already existing files and take actions based on what we found in the directory: rename existing 1, 2, 3-suffixed files so that we can use the "1" prefix for the currently rotated file, delete the file that falls out of the `rotate(N)` range, etc.

An in-memory counter is not enough here, as we'd accidentally override existing files after restarting syslog-ng.

(I'm more than happy if you're willing to work on it, but I'd like to mention that it's not a requirement for the project.)

shikharvashistha replied:
Have implemented [here](https://github.com/MrAnno/syslog-ng/blob/51bb5fed0c91897e69ea47912fefbe16df9a482b/modules/file-rotation/file-rotation.c#L74) kindly have a look

MrAnno replied:
Thank you, it is starting to take shape. `number_of_time_rotated` is still a memory-only value, this feature should work properly between restarts too, which requires checking already existing rotated files on the disk.

shikharvashistha replied:
I think [this](https://github.com/MrAnno/syslog-ng/blob/51bb5fed0c91897e69ea47912fefbe16df9a482b/modules/file-rotation/file-rotation.c#L59) will take care of that issue.

shikharvashistha committed:
file-rotation: file_rotation_set_number_of_file_rotatations

This commit introduces a new option to set number of file rotations possible and few validation checks

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha commented:
Hi @MrAnno, can we consider winding up the project?

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -135,7 +142,17 @@ log_proto_file_writer_flush(LogProtoClient *s)
+  const gchar *file_name = self->filename;
+  struct stat st;
+  //gsize file_size = lseek(self->fd, 0, SEEK_END);
+  stat(file_name, &st);
```
As we discussed, please remove the unconditional and expensive call from this side, and move it to the log rotation side.

`size` can be removed completely from `FileFlushSignalData`.

shikharvashistha replied:
Removed the occurrences of size from `FIleFlushSignalData`

MrAnno reviewed (commented):
Thank you.

Please consider fixing the above additional review notes (or reverting back to the non-`rotate()` version of the PR.

After that, please open a pull request against `syslog-ng/syslog-ng`.
Please don't forget to read our contribution guide, especially the section about pull requests:
https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#pull-requests

A few rounds of review will be done there too.
That may slip past the project timeline, but I hope you will still be available from time to time.

shikharvashistha committed:
file-rotation: Added checking file

This commit introduces checking file everytime on disk

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha commented:
> Thank you.
> 
> Please consider fixing the above additional review notes (or reverting back to the non-`rotate()` version of the PR.
> 
> After that, please open a pull request against `syslog-ng/syslog-ng`. Please don't forget to read our contribution guide, especially the section about pull requests: https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#pull-requests
> 
> A few rounds of review will be done there too. That may slip past the project timeline, but I hope you will still be available from time to time.

Yeah I'll be available for the review
